### DISCUSSION TITLE
[Upgrading problem] Failed to upgrade youtube-dl from 2021.4.7 to 2021.4.17 in brew

### DISCUSSION THREAD

**Discussion by EmmW** [Homebrew/discussions#1349] [Getting started]
System environment: Mac OS X 10.14.6

So in Terminal I have just run "brew update && brew upgrade" for a general upgrade of all brew components/extensions. After that I ran "youtube-dl --version" to double-check and surprisingly got back "2021.4.07". When I checked the parser it clearly appears that youtube-dl was indeed successfully upgraded:

==> Upgrading youtube-dl 2021.4.7 -> 2021.4.17
==> Downloading https://ghcr.io/v2/homebrew/core/youtube-dl/manifests/2021.4.17
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/youtube-dl/blobs/sha256:5bb2b85c53a5519077df1a9d2efd2bed1a4b52259e5ffc4aa6054de2f8383b73
==> Downloading from https://pkg-containers-az.githubusercontent.com/ghcr1/blobs/sha256:5bb2b85c53a5519077df1a9d2efd2bed1a4b52259e5ffc4aa6054de2f8383b73?se=2021-04-24T12%3A55%3A00Z&sig=BhxH
######################################################################## 100.0%
==> Pouring youtube-dl--2021.4.17.mojave.bottle.tar.gz
==> Caveats
Bash completion has been installed to:
/usr/local/etc/bash_completion.d
==> Summary
🍺 /usr/local/Cellar/youtube-dl/2021.4.17: 1,395 files, 13.5MB
Removing: /usr/local/Cellar/youtube-dl/2021.4.7... (1,394 files, 15.2MB)
Removing: /Users/myname/Library/Caches/Homebrew/youtube-dl--2021.4.7... (3.5MB)

I therefore tried again with "brew upgrade youtube-dl" and then got back "Warning: youtube-dl 2021.4.17 already installed". Another "youtube-dl --version" gave the same result: "2021.04.07"

Hence my question: What is wrong in my brew/youtube-dl configuration, and how do I go about to sort this out via Terminal?

**Comment by SMillerDev** [comment]
What does `brew info` say?

**Comment by carlocab** [comment]
What's the output of
```
brew info youtube-dl
readlink $(which youtube-dl)
```
?

**Comment by EmmW** [comment]
Thank you both for responding!

"brew info" gives:

113 kegs, 49,776 files, 2GB

"brew info youtube-dl" gives:

youtube-dl: stable 2021.4.17 (bottled)
Download YouTube videos from the command-line
https://youtube-dl.org/
/usr/local/Cellar/youtube-dl/2021.4.17 (1,395 files, 13.5MB) *
  Poured from bottle on 2021-04-24 at 14:47:50
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/youtube-dl.rb
License: Unlicense
==> Dependencies
Required: python@3.9 ✔
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d
==> Analytics
install: 112,994 (30 days), 305,015 (90 days), 1,030,831 (365 days)
install-on-request: 105,590 (30 days), 284,064 (90 days), 934,586 (365 days)
build-error: 0 (30 days)

"readlink $(which youtube-dl)" gives:

Illegal variable name.

"which youtube-dl" gives:

/Library/Frameworks/Python.framework/Versions/3.5/bin/youtube-dl

It would seem that the active youtube-dl (older v. 2021.04.7) resides within the older python3.5 folder, even though I somewhere else do have both python3.9.4 and youtube-dl 2021.04.17 installed..?

**Reply by carlocab** [reply]
You probably installed `youtube-dl` using a `pip3` that came with your older version of Python. What's the output of
```
which -a youtube-dl
which -a pip3
```
?
Do you remember doing a
```
sudo pip3 install youtube-dl
```
?

**Comment by EmmW** [comment]
When I run "which -a youtube-dl" I get:

-a: Command not found.
/Library/Frameworks/Python.framework/Versions/3.5/bin/youtube-dl

When I run "which -a pip3" I get:

-a: Command not found.
/Library/Frameworks/Python.framework/Versions/3.5/bin/pip3

I know I have in the past used both "brew install", "pip3 install" and "git pull" to get various versions of youtube-dl and svtplay-dl to work, so we are probably on to something here... Strangely enough, I used to have the exact same problem with svtplay-dl; brew would update it to the latest version, but Terminal would still use the older version (unless I managed to do some "git pull" procedure instead), but now brew and svtplay-dl works properly and instead I now seem to have gotten stuck with an old youtube-dl installation, even though brew updates it properly.

**Reply by carlocab** [reply]
What shell are you using? You shouldn't be seeing `-a: Command not found.` with `zsh` or `bash`.

In any case, you're stuck on the old version of `youtube-dl` because 
```
/Library/Frameworks/Python.framework/Versions/3.5/bin
```
is still in your `PATH`. (See also your [previous discussion post](https://github.com/Homebrew/discussions/discussions/1294).) The solution here would be to remove it or otherwise make sure that `/usr/local/bin` comes earlier in your `PATH`.

Otherwise, I'd try `pip3 uninstall youtube-dl` to try to get rid of the one that isn't installed by Homebrew.

**Comment by EmmW** [comment]
Thank you, problem solved (I figured it would be an easy fix if I got the answer from someone really knowledgeable about this :-)!

I wrote "sudo pip3 uninstall youtube-dl" with the following result:
```console
DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.
WARNING: The directory '/Users/markwinelid/Library/Caches/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Found existing installation: youtube-dl 2021.4.7
Uninstalling youtube-dl-2021.4.7:
  Would remove:
    /Library/Frameworks/Python.framework/Versions/3.5/bin/youtube-dl
    /Library/Frameworks/Python.framework/Versions/3.5/etc/bash_completion.d/youtube-dl.bash-completion
    /Library/Frameworks/Python.framework/Versions/3.5/etc/fish/completions/youtube-dl.fish
    /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/youtube_dl-2021.4.7.dist-info/*
    /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/youtube_dl/*
    /Library/Frameworks/Python.framework/Versions/3.5/share/doc/youtube_dl/README.txt
    /Library/Frameworks/Python.framework/Versions/3.5/share/man/man1/youtube-dl.1
Proceed (y/n)? y
  Successfully uninstalled youtube-dl-2021.4.7
```
After that I wrote "youtube-dl --version" and (finally!) got back the following: "2021.04.17", i.e. without having to re-address anything, brew now uses the latest version of youtube-dl I already had installed. Do you think running "sudo pip3 uninstall python3.5" could work to the same effect, so that brew would henceforth use my already installed python3.9.4 version instead of the old python3.5..?

**Reply by Bo98** [reply]
It looks like Python 3.5 was not installed via brew, so you'll need to read the Python docs on how to uninstall Python 3.5.

**Reply by carlocab** [reply]
You can only `pip3 uninstall` something you `pip3 install`-ed, and I don't think you `pip3 install`-ed Python 3.5. The solution you need is one I've already mentioned: https://github.com/Homebrew/discussions/discussions/1294#discussioncomment-620203
